src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsWriter.java

```diff
@@ -245,12 +248,22 @@ private <T, C> void trainAndIndex(
             : NativeIndexWriter.getWriter(fieldInfo, segmentWriteState);
 
         knnVectorValues = vectorValuesRetriever.apply(vectorDataType, fieldInfo, VectorProcessingContext);
-
+        if (shouldSkipBuildingVectorDataStructure(knnVectorValues.totalLiveDocs())) {
+            log.info("Skip building vector data structure for field: ", fieldInfo.name);
+            return;
+        }
         StopWatch stopWatch = new StopWatch();
         stopWatch.start();
         indexOperation.buildAndWrite(writer, knnVectorValues);
         long time_in_millis = stopWatch.totalTime().millis();
         graphBuildTime.incrementBy(time_in_millis);
         log.warn("Graph build took " + time_in_millis + " ms for " + operationName);
     }
+
+    private boolean shouldSkipBuildingVectorDataStructure(final long docCount) {
```

Comment:
@shatejas Are you suggesting that it should always return false on merge? If so, not necessarily, we don't want to build graph during background merges as well.

Reply:
I missed the background merge case. A few questions around background merges
1. When does it trigger?
2. is there a total indexing time latency if we allow it during background merge? if so by how much
3. Is there a CPU impact if we allow background merges? whats the ballpark estimates for that

Reply:
OpenSearch uses TieredMergePolicy to decide when to perform merge. IIRC, segmentsPerTier controls how many segments are allowed per tiers. When segment count goes beyond that limit in tier, it will run merges in the background. Choosing this value is generally a trade-off of indexing speed vs searching speed. Having fewer segments improves search since lucene need to look into less number of segments, but, this will increasing indexing time.

Reply:
Thanks for the link @VijayanB

1.  I am trying to grok why merges are behind this setting. and thus the questions on indexing time and CPU usage. Keeping background merges enabled overall seems to better the user experience in few ways (Please point out if I am missing something or incorrect)
- Users can have a set and forget setting - once the user sets 0 or a threshold value they don't have to change it for force merges, and then change it back when they want to index again. 
- Better search experience - Background merges may create some graphs, which will lower the latencies if users chose to search during indexing as it won't always hit exact search
- Force merge optional - Since background merges will create graphs, guidance on force merge can be optional/enforced only for cases where threshold is low/0. 

2. Can the default threshold be kept 0 if merge is kept enabled and the trigger for merge is predictable? This is just a thought to see if its possible as this will give the max total indexing time benefit for all users.

Reply:
> I am trying to grok why merges are behind this setting. and thus the questions on indexing time and CPU usage. Keeping background merges enabled overall seems to better the user experience in few ways

No it is correct experience. It just defeats the overall purpose why the change at first place is getting added. 

> Users can have a set and forget setting - once the user sets 0 or a threshold value they don't have to change it for force merges, and then change it back when they want to index again.

This is what threshold based graph creation will do. If user want to just do the graph creation during merges they can keep the threshold high.

> Better search experience - Background merges may create some graphs, which will lower the latencies if users chose to search during indexing as it won't always hit exact search

If user is looking to do so then its better to use the threshold. 


> Force merge optional - Since background merges will create graphs, guidance on force merge can be optional/enforced only for cases where threshold is low/0.

No it is not optional. Because even when you have small segments its not necessary that merges will happen. 

> Can the default threshold be kept 0 if merge is kept enabled and the trigger for merge is predictable?

Trigger for merge is very much dependent on indexing speed. You might not have merges for 1 whole day but the moment you add 1 more doc in index the merges may get triggered, so it really depends on the state of the shard aka Lucene index.

Since there is no formal documentation or wiki around merges I would recommend reading this blog : https://blog.mikemccandless.com/2011/02/visualizing-lucenes-segment-merges.html which is an old blog from one of lucene maintainer which might help resolve some of the doubts you have related to merges.

Reply:
Understood, so seems like we need to the user to change the setting for force merge

Reply:
Yes.

Reply:
@shatejas Added unit test by rebasing from main.